master/internal/configpolicy/utils.go

```diff
@@ -95,10 +120,18 @@ func ValidateExperimentConfig(
 	if cp.InvariantConfig != nil {
 		if cp.InvariantConfig.RawResources != nil {
 			checkAgainstGlobalPriority(priorityEnabledErr, cp.InvariantConfig.RawResources.RawPriority)
+
+			// Verify the workspace invariant config doesn't conflict with workspace constraints.
 			if err := checkConstraintConflicts(cp.Constraints, cp.InvariantConfig.RawResources.RawMaxSlots,
 				cp.InvariantConfig.RawResources.RawSlotsPerTrial, cp.InvariantConfig.RawResources.RawPriority); err != nil {
 				return status.Errorf(codes.InvalidArgument, fmt.Sprintf(InvalidExperimentConfigPolicyErr+": %s.", err))
 			}
+
+			// Verify the workspace invariant config doesn't conflict with global constraints.
+			if err := checkConstraintConflicts(globalConstraints, cp.InvariantConfig.RawResources.RawMaxSlots,
```

Comment:
I think this code block fixes the logic hole. The other changes confuse "what's different" and make it harder to understand what's changed.

Reply:
I've done some function abstraction to make the changes clearer!